### PR TITLE
Document pitfall with `impl PartialEq<B> for A`

### LINKED ISSUE #66476: Documentation of PartialEq has example that violates transitivity
The [Documentation of `PartialEq`](https://doc.rust-lang.org/std/cmp/trait.PartialEq.html) states that implementations must be transitive: for all for all `a`, `b` and `c`, `a == b` and `b == c` implies `a == c`. This is followed by a note indicating that the transitivity must also hold if `a`, `b`, and `c` are of different types.

However, the last example implementation in the documentation is not transitive:
```rust
let b1 = Book { isbn: 1, format: BookFormat::Paperback };
let b2 = Book { isbn: 2, format: BookFormat::Paperback };

assert!(b1 == BookFormat::Paperback);
assert!(BookFormat::Paperback == b2);

// The following should hold by transitivity but doesn't.
assert!(b1 == b2); // <-- PANICS
```

I can think of three possible ways to fix this:
- Remove the last example.
- Keep the last example as an explicit counterexample to warn how easy it is to accidentally write implementations that violate the contract.
- Modify the last example in a way that fixes the issue (I'm not sure what that would look like).

I can submit a pull request if I get some guidance which of the above solutions to choose (I personally have a weak preference for the second one).

### PR DESCRIPTION
**Pull request by robamler**
Fixes #66476 by turning the violating example into an explicit
counterexample.

### CODE REVIEW

**Comment by robamler** [comment]
Sorry for the two separate commits. I forgot to declare the counterexample as `should_panic`. Please let me know if I should squash the two commits into one and then force-push.

File: src/libcore/cmp.rs
Comment (rkruppe) [inline src/libcore/cmp.rs]: Typo
```suggestion
/// requirements for a partial equivalence relation. For example, if we kept
```

**Review by hanna-kruppe — APPROVED** [review]
LGTM (except for a typo), thanks!

Separate commits are generally good for review, we just squash them after review / before merging. Which in this case would be after fixing the typo, then I'll happily merge this.

**Comment by robamler** [comment]
Thank you for fixing the typo, Dylan!

@rkruppe Do you need me to squash or do you use the Github "Squash & Merge" feature? (Sorry, couldn't find this in the the [contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md).)

**Comment by rkruppe** [comment]
We're not using any variant of Github's merge button, it all goes through our integration bot @bors. They can't squash, so it'll have to happen manually. Sorry for the confusion.

**Comment by robamler** [comment]
Squashed it into a single commit, should be ready to merge.

**Comment by rkruppe** [comment]
Thanks!

@bors r+ rollup

**Pull request merged**
